### PR TITLE
Timing Calibrations for Totem Timing detectors

### PR DESCRIPTION
**Pull request by nminafra**
Code written by @pilifjed to calibrate ADC non-linearity and time offset per channel.
@forthommel, do you have any suggestion on the json part?

### CODE REVIEW

File: RecoCTPPS/TotemRPLocal/interface/CalibrationData.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#ifndef RecoCTPPS_TotemRPLocal_JsonRead
```
Comment (nminafra) [inline RecoCTPPS/TotemRPLocal/interface/CalibrationData.h]: Should this be deleted? there is TotemTimingParser

File: RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h [hunk 2]
```diff
@@ -12,33 +12,34 @@
...
-
-  const float getTriggerTime(const TotemTimingDigi& digi) const;
-
-  std::vector<float> getTimeSamples(const TotemTimingDigi& digi) const;
-
-  std::vector<float> getVoltSamples(const TotemTimingDigi& digi) const;
+  TotemTimingConversions(bool mergeTimePeaks);
+  TotemTimingConversions(bool mergeTimePeaks, const std::string& calibrationFile);
+  void openCalibrationFile();
+  void openCalibrationFile(const std::string& calibrationFile);
```
[the remark this answers is not in the document]
Comment (forthommel) [reply RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h]: Hi! my 2 cents: maybe `edm::FileInPath` is more convenient here...

File: RecoCTPPS/TotemRPLocal/src/TotemTimingRecHitProducerAlgorithm.cc [hunk 3]
```diff
@@ -17,7 +17,7 @@ const float TotemTimingRecHitProducerAlgorithm::SINC_COEFFICIENT = std::acos(-1)
 
 TotemTimingRecHitProducerAlgorithm::TotemTimingRecHitProducerAlgorithm(
     const edm::ParameterSet &iConfig)
-    : sampicConversions_(iConfig.getParameter<std::string>("calibrationFile")),
+    : sampicConversions_(TotemTimingConversions(iConfig.getParameter<bool>("mergeTimePeaks"), iConfig.getParameter<std::string>("calibrationFile"))),
```
Comment (forthommel) [inline RecoCTPPS/TotemRPLocal/src/TotemTimingRecHitProducerAlgorithm.cc]: Constructor + copy; I would suggest you to stick to the previous version with an extra argument

File: RecoCTPPS/TotemRPLocal/src/TotemTimingParser.cc [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  return formula_;
+}
+
+std::ostream& operator<<(std::ostream &strm, const TotemTimingParser::CalibrationKey &key){
+  return strm << key.db << " " << key.sampic << " " << key.channel << " " << key.cell;
+}
+
+std::ostream& operator<<(std::ostream &strm, const TotemTimingParser &data){
+  strm << "FORMULA: "<< data.formula_ << "\nDB SAMPIC CHANNEL CELL PARAMETERS TIME_OFFSET\n";
+  for (auto& kv : data.parameters_) {
```
Comment (forthommel) [inline RecoCTPPS/TotemRPLocal/src/TotemTimingParser.cc]: Use `const` when possible

File: RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h [hunk 5]
```diff
@@ -12,33 +12,34 @@
...
-  void openCalibrationFile(const std::string& calibrationFile="/dev/null");
-
-  const float getTimeOfFirstSample(const TotemTimingDigi& digi) const;
-
-  const float getTriggerTime(const TotemTimingDigi& digi) const;
-
-  std::vector<float> getTimeSamples(const TotemTimingDigi& digi) const;
-
-  std::vector<float> getVoltSamples(const TotemTimingDigi& digi) const;
+  TotemTimingConversions(bool mergeTimePeaks);
```
[the remark this answers is not in the document]
Comment (pilifjed) [reply RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h]: done 09c7da5

File: RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (pilifjed) [reply RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h]: overloading done 09c7da5

[the remark this answers is not in the document]
Comment (nminafra) [reply RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h]: Reference to the FileInPath:
http://cmsdoxygen.web.cern.ch/cmsdoxygen/CMSSW_10_3_0_pre2/doc/html/d6/d01/classedm_1_1FileInPath.html
@forthommel maybe we can discuss this in person

**Pull request merged**